File: Task/Hamiltonian-Cycle-via-Chv-tal-Closure/R/hamiltonian-cycle-via-chv-tal-closure.r
# Graph structure as a list with n (number of vertices) and adj (adjacency matrix)
Graph <- function(n, adj = NULL) {
  if (is.null(adj)) {
    adj <- matrix(FALSE, nrow = n, ncol = n)
  }
  list(n = n, adj = adj)
}

# Add an undirected edge u--v
add_edge <- function(g, u, v) {
  if (0 < u && u <= g$n && 0 < v && v <= g$n) {
    g$adj[u, v] <- TRUE
    g$adj[v, u] <- TRUE
  }
  g
}

# Degree of vertex u
degree <- function(g, u) {
  sum(g$adj[u, ])
}

# Compute the Chvátal closure in-place
closure <- function(g) {
  n <- g$n
  repeat {
    added <- FALSE
    # Try every non-edge (u,v) with u < v
    found <- FALSE
    for (u in 1:(n-1)) {
      for (v in (u+1):n) {
        if (!g$adj[u, v]) {
          du <- degree(g, u)
          dv <- degree(g, v)
          if (du + dv > n) {
            # add the edge
            g <- add_edge(g, u, v)
            added <- TRUE
            found <- TRUE
            break
          }
        }
      }
      if (found) break
    }
    if (!added) break
  }
  g
}

# Is the graph complete?
is_complete <- function(g) {
  for (u in 1:(g$n-1)) {
    for (v in (u+1):g$n) {
      if (!g$adj[u, v]) {
        return(FALSE)
      }
    }
  }
  TRUE
}

# Depth-first search for Hamiltonian cycle
dfs <- function(g, u, visited, path) {
  if (length(path) == g$n) {
    # check if can close the cycle
    if (g$adj[u, path[1]]) {
      return(c(path, path[1]))
    } else {
      return(NULL)
    }
  }

  for (v in 1:g$n) {
    if (!visited[v] && g$adj[u, v]) {
      visited[v] <- TRUE
      path <- c(path, v)
      cycle <- dfs(g, v, visited, path)
      if (!is.null(cycle)) {
        return(cycle)
      }
      # else backtrack if no cycle yet
      path <- path[-length(path)]
      visited[v] <- FALSE
    }
  }
  NULL
}

# Find a Hamiltonian cycle in the original graph (not closure)
# by simple backtracking. Returns vertices in order (1..n), and back to start.
hamiltonian_cycle <- function(g) {
  visited <- rep(FALSE, g$n)
  # set starting vertex
  path <- 1
  visited[1] <- TRUE
  dfs(g, 1, visited, path)
}

test_hamiltonian <- function() {
  # Example: 5 vertices, almost complete graph missing edge 1--2.
  # This satisfies Ore's condition: any non-edge (1,2) has deg(1)=3, deg(2)=3, 3+3>=5.
  g <- Graph(5)

  # Add all edges except (1,2)
  for (u in 1:4) {
    for (v in (u+1):5) {
      if (!(u == 1 && v == 2)) {
        g <- add_edge(g, u, v)
      }
    }
  }

  cat("Original graph degrees:\n")
  for (u in 1:g$n) {
    cat(sprintf(" deg(%d) = %d\n", u, degree(g, u)))
  }

  # Compute closure (create a copy first)
  c <- Graph(g$n, g$adj)
  c <- closure(c)

  cat("\nAfter Chvátal closure (translated to 0-based indices):\n")
  for (u in 1:c$n) {
    cat(u - 1, ": ")
    for (v in 1:c$n) {
      if (c$adj[u, v]) {
        cat(v - 1, " ")
      }
    }
    cat("\n")
  }

  if (is_complete(c)) {
    cat("\nClosure is complete ⇒ graph is Hamiltonian (by Bondy–Chvátal).\n")
    cycle <- hamiltonian_cycle(g)
    if (!is.null(cycle)) {
      cat("Found Hamiltonian cycle in original graph, as zero-based is:\n")
      for (i in seq_along(cycle)) {
        if (i > 1) cat(" → ")
        cat(cycle[i] - 1)
      }
      cat("\n")
    } else {
      cat("Unexpected: could not find a Hamiltonian cycle.\n")
    }
  } else {
    cat("\nClosure is not complete ⇒ no guarantee of Hamiltonian cycle.\n")
  }
}

# Run the test
test_hamiltonian()

File: Task/Hamiltonian-Cycle-via-Chv-tal-Closure/Julia/hamiltonian-cycle-via-chv-tal-closure.jl
struct Graph
    n::Int
    adj::Matrix{Bool}
end
Graph(n) = Graph(n, falses(n, n))

""" Add an undirected edge u--v. """
function add_edge!(g::Graph, u, v)
    if 0 < u <= g.n && 0 < v <= g.n
        g.adj[u, v] = true
        g.adj[v, u] = true
    end
end

""" Degree of vertex u. """
degree(g::Graph, u) = count(@view g.adj[u, :])

""" Compute the Chvátal closure in‑place. """
function closure!(g::Graph)
    n = g.n
    while true
        added = false
        # Try every non-edge (u,v) with u < v
        for u in 1:n-1, v in u+1:n
            if !g.adj[u, v]
                du = degree(g, u)
                dv = degree(g, v)
                if du + dv > n
                    # add the edge
                    add_edge!(g, u, v)
                    added = true
                    break
                end
            end
        end
        !added && break
    end
end

""" Is the graph complete? """
function is_complete(g::Graph)
    for u in 1:g.n, v in u+1:g.n
        !g.adj[u, v] && return false
    end
    return true
end

function dfs(g::Graph, u, visited, path)
    if length(path) == g.n
        # check if can close the cycle
        return g.adj[u, path[begin]] ? [path; path[begin]] : nothing
    end
    for v in 1:g.n
        if !visited[v] && g.adj[u, v]
            visited[v] = true
            push!(path, v)
            cycle = dfs(g, v, visited, path)
            !isnothing(cycle) && return cycle
            # else backtrack if no cycle yet
            pop!(path)
            visited[v] = false
        end
    end
    return nothing
end

""" Find a Hamiltonian cycle in the original graph (not closure)
    by simple backtracking. Returns vertices in order (0..n-1), and back to start.
"""
function hamiltonian_cycle(g::Graph)
    visited = falses(g.n)
    # set starting vertex
    path = [1]
    visited[begin] = true
    return dfs(g, 1, visited, path)
end

function test_hamiltonian()
    # Example: 5 vertices, almost complete graph missing edge 0--1.
    # This satisfies Ore's condition: any non-edge (0,1) has deg(0)=3, deg(1)=3, 3+3>=5.
    g = Graph(5)
    # Add all edges except (0,1)
    for u in 1:4
        for v in u+1:5
            if !(u == 1 && v == 2)
                add_edge!(g, u, v)
            end
        end
    end

    println("Original graph degrees:")
    for u in 1:g.n
        println(" deg($u) = ", degree(g, u))
    end

    # Compute closure
    c = Graph(g.n, copy(g.adj))
    closure!(c)

    println("\nAfter Chvátal closure (translated to 0-based indices):")
    for u in 1:c.n
        print(u - 1, ": ")
        for v in 1:c.n
            if c.adj[u, v]
                print(v - 1, " ")
            end
        end
        println()
    end

    if is_complete(c)
        println("\nClosure is complete ⇒ graph is Hamiltonian (by Bondy–Chvátal).")
        cycle = hamiltonian_cycle(g)
        if !isnothing(cycle)
            println("Found Hamiltonian cycle in original graph, as zero-based is:")
            for (i, v) in enumerate(cycle)
                i > 1 && print(" → ")
                print(v - 1)
            end
            println()
        else
            println("Unexpected: could not find a Hamiltonian cycle.")
        end
    else
        println("\nClosure is not complete ⇒ no guarantee of Hamiltonian cycle.")
    end
end

test_hamiltonian()
